src/tty/windows.rs

```diff
@@ -62,111 +50,411 @@ pub fn process_should_exit() -> bool {
+
+                assert!(success);
+            }
+
+            // Get process commandline
+            let default_shell = &Shell::new(env::var("COMSPEC").unwrap_or_else(|_| "cmd".into()));
+            let shell = config.shell().unwrap_or(default_shell);
+            let initial_command = options.command().unwrap_or(shell);
+            let mut cmdline = initial_command.args().to_vec();
+            cmdline.insert(0, initial_command.program().into());
+
+            // Warning, here be borrow hell
+            let cwd = options.working_dir.as_ref().map(|dir| canonicalize(dir).unwrap());
+            let cwd = cwd.as_ref().map(|dir| dir.to_str().unwrap());
+
+            // Create the client application, using startup info containing ConPTY info
+            let cmdline = U16CString::from_str(&cmdline.join(" ")).unwrap().into_raw();
+            let cwd = cwd.map(|s| U16CString::from_str(&s).unwrap());
+            let cwd_ptr = match &cwd {
+                Some(b) => b.as_ptr() as LPCWSTR,
+                None => ptr::null()
+            };
+
+            let mut proc_info: PROCESS_INFORMATION = Default::default();
+            unsafe {
+                success = CreateProcessW(
+                    ptr::null(),
+                    cmdline as LPWSTR,
+                    ptr::null_mut(),
+                    ptr::null_mut(),
+                    true as i32,
+                    EXTENDED_STARTUPINFO_PRESENT,
+                    ptr::null_mut(),
+                    cwd_ptr,
+                    &mut startup_info_ex.StartupInfo as *mut STARTUPINFOW,
+                    &mut proc_info as *mut PROCESS_INFORMATION
+                ) > 0;
+
+                assert!(success);
+            }
+
+            // Recover raw memory to cmdline so it can be freed
+            unsafe { U16CString::from_raw(cmdline); }
+
+            // Store handle to console
+            unsafe { HANDLE = proc_info.hProcess; }
+
+            let conin = AsyncAnonWrite::new(conin);
+            let conout = AsyncAnonRead::new(conout);
+
+            let agent = Conpty {
+                handle: pty_handle
+            };
+
+            (agent, conout, conin)
+        })
+    }
+
+    impl OnResize for HPCON {
+        fn on_resize(&mut self, sizeinfo: &SizeInfo) {
+            if let Some(api) = &*CONPTY_API {
+                let result = unsafe {
+                    (api.ResizePseudoConsole)(
+                        *self,
+                        COORD {X: sizeinfo.cols().0 as _, Y: sizeinfo.lines().0 as _}
+                    )
+                };
+                assert!(result == S_OK);
+            }
+        }
+    }
+}
+
+pub trait EventedRead: Read + Evented { }
+impl<T: Read + Evented> EventedRead for T { }
+
+pub trait EventedWrite: Write + Evented { }
+impl<T: Write + Evented> EventedWrite for T { }
+
+pub enum Agent<'a> {
+    Winpty(winpty::Agent<'a>),
+    Conpty(conpty::Agent)
+}
+
+pub struct Pty<'a> {
+    agent: Arc<Mutex<Agent<'a>>>,
```

Comment:
I don't like having the mutex here. Winpty already uses one internally, though admittedly I did have troubles changing the bindings to reflect that (using `UnsafeCell` internally to change the receiver to `&self`).

Does the conpty API need the mutex?